sympy/matrices/tests/test_matrices.py

```diff
@@ -259,6 +265,26 @@ def test_power():
     raises(ValueError, lambda: An.subs(n, -2).doit())
     assert An * An == A**(2*n)
 
+    # concretizing behavior for non-integer and complex powers
+    A = Matrix([[0,0,0],[0,0,0],[0,0,0]])
+    n = Symbol('n', integer=True, positive=True)
+    assert A**n == A
+    n = Symbol('n', integer=True, nonnegative=True)
+    assert A**n == diag(0**n, 0**n, 0**n)
+    assert (A**n).subs(n, 0) == eye(3)
+    assert (A**n).subs(n, 1) == zeros(3)
+    A = Matrix ([[2,0,0],[0,2,0],[0,0,2]])
+    assert A**2.1 == diag (2**2.1, 2**2.1, 2**2.1)
+    assert A**I == diag (2**I, 2**I, 2**I)
+    A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])
+    raises(ValueError, lambda: A**2.1)
+    raises(ValueError, lambda: A**I)
+    A = Matrix([[S.Half, S.Half],[S.Half, S.Half]])
+    assert A**S.Half == A
+    A = Matrix([[1, 1],[3, 3]])
+    assert (A**S.Half)**2 == A
+    assert (A**2)**S.Half == A
```

Comment:
It would be better to explicitly test what `A**2` and `A**S.Half` give here. These tests would pass even if the power doesn't evaluate e.g.:
```julia
In [13]: A**n                                                                                                                     
Out[13]: 
        n
⎛⎡1  1⎤⎞ 
⎜⎢    ⎥⎟ 
⎝⎣3  3⎦⎠ 

In [14]: (A**n)**(1/n)                                                                                                            
Out[14]: 
⎡1  1⎤
⎢    ⎥
⎣3  3⎦
```